src/main/java/org/elasticsearch/script/ScriptService.java

```diff
@@ -217,60 +216,70 @@ private ScriptEngineService getScriptEngineServiceForFileExt(String fileExtensio
+        if (canExecuteScript(lang, scriptEngineService, script.type, script.context) == false) {
+            throw new ScriptException("scripts of type [" + script.type + "], operation [" + script.context.getKey() + "] and lang [" + lang + "] are disabled");
         }
-        return compileInternal(lang, script, scriptType);
+
+        return compileInternal(script);
     }
 
     /**
      * Compiles a script straight-away, or returns the previously compiled and cached script, without checking if it can be executed based on settings.
      */
-    public CompiledScript compileInternal(String lang,  String script, ScriptType scriptType) {
-        assert script != null;
-        assert scriptType != null;
+    public Script compileInternal(Script script) {
+        assert script.key != null;
+        assert script.type != null;
+
+        String lang = script.lang;
+
         if (lang == null) {
             lang = defaultLang;
         }
         if (logger.isTraceEnabled()) {
-            logger.trace("Compiling lang: [{}] type: [{}] script: {}", lang, scriptType, script);
+            logger.trace("Compiling lang: [{}] type: [{}] script: {}", lang, script.type, script.key);
         }
 
         ScriptEngineService scriptEngineService = getScriptEngineServiceForLang(lang);
-        CacheKey cacheKey = newCacheKey(scriptEngineService, script);
+        CacheKey cacheKey = newCacheKey(scriptEngineService, script.key);
 
-        if (scriptType == ScriptType.FILE) {
-            CompiledScript compiled = staticCache.get(cacheKey); //On disk scripts will be loaded into the staticCache by the listener
+        if (script.type == ScriptType.FILE) {
+            Object compiled = staticCache.get(cacheKey); //On disk scripts will be loaded into the staticCache by the listener
             if (compiled == null) {
-                throw new ElasticsearchIllegalArgumentException("Unable to find on disk script " + script);
+                throw new ElasticsearchIllegalArgumentException("Unable to find on disk script " + script.key);
             }
-            return compiled;
+            return new Script(script, lang, compiled);
         }
 
-        if (scriptType == ScriptType.INDEXED) {
-            final IndexedScript indexedScript = new IndexedScript(lang, script);
-            script = getScriptFromIndex(indexedScript.lang, indexedScript.id);
-        }
+        Object compiled = cache.getIfPresent(cacheKey);
 
-        CompiledScript compiled = cache.getIfPresent(cacheKey);
         if (compiled == null) {
+            String source = script.key;
+
+            if (script.type == ScriptType.INDEXED) {
+                final IndexedScript indexedScript = new IndexedScript(lang, script.key);
+                source = getScriptFromIndex(indexedScript.lang, indexedScript.id);
+            }
```

Comment:
Done.  I realized that caching for indexed scripts was broken, but I thought this might be a bit more performant at the time.  I definitely don't want to muck with a potential fix, though.